app/builders/bulk_subscriber_list_email_builder.rb

```diff
@@ -1,40 +1,81 @@
 class BulkSubscriberListEmailBuilder < ApplicationBuilder
+  BATCH_SIZE = 5000
+
   def initialize(subject:, body:, subscriber_lists:)
     @subject = subject
     @body = body
     @subscriber_lists = subscriber_lists
+    @now = Time.zone.now
   end
 
   def call
-    records.any? ? Email.insert_all!(records).pluck("id") : []
+    ActiveRecord::Base.transaction do
+      batches.flat_map do |batch|
+        records = records_for_batch(batch.to_h)
+        Email.insert_all!(records).pluck("id")
+      end
+    end
   end
 
 private
 
-  attr_reader :subject, :body, :subscriber_lists
-
-  def records
-    @records ||= begin
-      now = Time.zone.now
-      subscribers.map do |address, subscriber_id|
-        {
-          address: address,
-          subject: subject,
-          body: body,
-          subscriber_id: subscriber_id,
-          created_at: now,
-          updated_at: now,
-        }
-      end
+  attr_reader :subject, :body, :subscriber_lists, :now
+
+  def records_for_batch(batch)
+    subscribers = Subscriber.find(batch.keys).index_by(&:id)
```

Comment:
Feel free to ignore:
I'm not v familiar with email alert api. It took me a while to figure out that calling `.to_h` on the array produced by the scope `.subscription_ids_by_subscriber` would mean that `batch.key` was a subscriber id. and like wise `batch.values` are subscription ids. Perhaps `batch.keys` and `batch.values` could be their own more explicitly named methods?

Reply:
Yes, it's pretty obscure isn't it! Apparently it's [a Rails-ism](https://apidock.com/ruby/v2_5_5/Array/to_h), which I copied from [here](https://github.com/alphagov/email-alert-api/blob/master/app/services/immediate_email_generation_service.rb#L36). I'm hoping to make this clearer in the follow-up PR, but for now I've amended [the fourth commit](https://github.com/alphagov/email-alert-api/pull/1542/commits/c872fa49f54eebb2bb67f6054edd254ca2c27387) so it's hopefully a little clearer. 

How does it look now?